### DISCUSSION TITLE
Youtube Live to transcript

### DISCUSSION THREAD

**Discussion by satish860** [deepgram/community#255] [General help] [answered]
Hi All, I am developing software that can convert courtroom proceedings from speech to text. Currently, the proceedings are streamed using YouTube Live. I tried the YouTube live video and code below, but I am promptly disconnected.  Did not locate any error messages. Would like to comprehend If anyone has previously attempted this,

```
const fetch = require('cross-fetch');
const { Deepgram } = require('@deepgram/sdk');

// The API key you created in step 1
const deepgramApiKey = '<<API Token removed for security purposes>>';

// URL for the real-time streaming audio you would like to transcribe
const url = 'https://www.youtube.com/live/MLraIRibKgI?feature=share';

// Initialize the Deepgram SDK
const deepgram = new Deepgram(deepgramApiKey);

// Create a websocket connection to Deepgram
const deepgramLive = deepgram.transcription.live({
    punctuate: true, 
    model: 'nova', 
    language: 'en-IN'
});

// Listen for the connection to open and send streaming audio from the URL to Deepgram
fetch(url).then(r => r.body).then(res => {
  res.on('readable', () => {
    if(deepgramLive.getReadyState() === 1) {
      deepgramLive.send(res.read());
    }
  });
});

// Listen for the connection to close
deepgramLive.addListener('close', () => {
  console.log('Connection closed.')
});

// Listen for any transcripts received from Deepgram and write them to the console
deepgramLive.addListener('transcriptReceived', (transcription) => {
  console.dir(transcription, { depth: null });
});
```

**Comment by jjmaldonis** [comment]
Hey @satish860, this is a great idea. We just released our [youtube demo](https://youtube-demo.deepgram.com/) a week ago and we do not have support for live streaming youtube videos... yet! This is something we want to add, but can't give an ETA quite yet.

**Reply by satish860** [reply]
That is really great 👍. My application already supports youtube using deepgram but it doesn't support youtube live alone. It will be great if you point me to resources then I will also try from my end and as always I would love to give attributions for the same and pay for the same.

**Comment by jjmaldonis** [comment, marked as answer]
Hey @satish860 check out this code.

The code is written in Python rather than Node and the code relies on a package called [streamlink](https://github.com/streamlink/streamlink). Hopefully there is an equivalent package in Node.

This should be used as a proof-of-concept that this is even possible. You'll want to optimize the sleep time and data collection size. You may also want to choose a different audio stream, or be more intelligent about the way the stream is chosen.

The streamlink package supports many different services including both YouTube and Twitch.

```
import streamlink
import websockets
import asyncio
import json
import os


DEEPGRAM_API_KEY = os.environ["DEEPGRAM_API_KEY"]


def get_audio_stream(url: str):
    streams = streamlink.streams(url)
    return streams["audio_mp4a"]


async def stream(audio_stream):
    async with websockets.connect(
        "wss://api.deepgram.com/v1/listen",
        extra_headers={"Authorization": f"Token {DEEPGRAM_API_KEY}"},
    ) as ws:

        async def sender(ws):
            while True:
                data = audio_stream.read(4096)
                await asyncio.sleep(0.1)
                await ws.send(data)

        async def receiver(ws):
            async for msg in ws:
                data = json.loads(msg)
                print(data["channel"]["alternatives"][0]["transcript"])
            return

        functions = [
            asyncio.ensure_future(sender(ws)),
            asyncio.ensure_future(receiver(ws)),
        ]
        await asyncio.gather(*functions)


def main():
    url = "https://www.youtube.com/watch?v=fvdQ47HuSDo"
    audio_stream = get_audio_stream(url)
    new_loop = asyncio.new_event_loop()  # Hacky due to how streamlink is implemented
    new_loop.run_until_complete(stream(audio_stream.open()))


if __name__ == "__main__":
    main()
```

**Reply by liamb0t** [reply]
I'm curious how to get this to work with a Twitch stream.

Twitch doesn't have audio_mp4a, streamlink returns a URL with a .m3u8 extension on the end.

If I drop it in VLC media it streams the audio fine but if I put it in Deepgram the program closes instantly with a code 1000.

{'audio_only': <TwitchHLSStream ['hls', 'https://video-weaver.sel04.hls.ttvnw.net/v1/playlist/CskEx5lTHs6u66QyN11E7lMm5D-LvfHcf4hXvWWiYJ3eRWI50qDMESgJ_d2ohv7ubS9hRmWXYbCp_Hltp2lcuaDi7yZDU0FRdCOHVsMctYiH1mMdNtv4n7CsOF1hKyn5v3Okq1xrZ7G1CtlyP6IXHBhzG0nEJKKd2cGWWJWx3PPsbLdFRoG7XvkbnpGiyh7TwwpKjF38hXUGoqTDhX7-dQ_f3Kq2w4BE7jmpv2JgUCuxj_pIfPSJgSJsyJ-X08XOjfdV_b9aCzmpu6ERBAKYOXKCmWiSuAnkM7IKq7VydhVecFZcdWoTqXuOOUZ3gcCvlEc3g2Kwm4GupdCz3pq24lHmo1_ma7xOvVQpWrQOxU4wMonjGFMQlci2eS_KxiHncBwHzTzIxtur37213ydoTJpIB9a6JRNNp8bqqm8hCzYduPzyITvgi1gy4uoxs59fYjK0fB6_-MvSeRTLFjV6JFLcsKj01khgfIbeHzI5Xna2qgkgHCuntgy_gVQqCyUp9yNsXlEynWJ9CTewt1r9QN6_J3a6aR1hZV_2XXau-lsyEPKLBM8JtbQMP5sWPeVB8n835q7Hu7bDjmljlkEsIBZtqsvQllhnAidd1MPzAgICIk9OCY-_Gjs18agKivByZxfQGxfSRphJX2eKt3uxLqbKl_T-c_155e7oeCqjhYbzANzmpMr2o46gdarPiWT0EST3zDGTkrD59iFUXhE2SeWk3XRlSzVxWpreWYtv-G4QJtaiHRr6MlTbVdkJVxJth9A5TTM6LI4X-i5hGgzGVIYF60RsVXqLgeIgASoJdXMtZWFzdC0yMIcI.m3u8'
